new getNodeForPopup method and more

daviddbal opened the pull request:
Tom,

This PR contains several commits that do the following:
- adds selectOneAppointmentCallback that is run when only one appointment is selected
- renames the callbacks I added to follow the verb-noun convention like the other callbacks in Agenda.
- adds -fx-fill colors to Agenda.css to allow me to use Rectangles as icons.
- adds getNodeForPopup and a map to back it in AgendaSkinTimeScale24HourAbstract.

The appointmentBodyPaneMap I added in the second commit has been removed.

-David

daviddbal committed:
new select one appointment callback

daviddbal committed:
new select one appointment callback

daviddbal committed:
added appointmentBodyPaneMap to match Appointments to its body pane

daviddbal committed:
added appointmentBodyPaneMap to match Appointments to its body pane

tbee committed:
added appointmentChangedCallback

daviddbal committed:
callback name changes to match pattern of older callbacks

daviddbal committed:
Added -fx-fill colors to the css group style class names

I want to use Rectangles as graphic icons instead of Panes.  I need -fx-
fill for the Rectangles.

daviddbal committed:
added getNodeForPopup to AgendaSkin and map to back it in AgendaSkinTimeScale24HourAbstract

daviddbal committed:
added getNodeForPopup to AgendaSkin and map to back it in AgendaSkinTimeScale24HourAbstract

daviddbal committed:
removed unused imports

daviddbal committed:
removed unused imports

tbee commented:
Where do you need the oneAppointmentCallback for? De appointments observable collection should provide the same info?

daviddbal commented:
That's a good point.  I can listen to changes to selectedAppointments, if it contains only one then I can do what ever I want.  Is that what you mean?

tbee commented:
yeah

daviddbal committed:
removed selectOneAppointmentCallback

replaced with selectedAppointment appointment listener outside Agenda

daviddbal commented:
I made the change.  The selecte one appointment callback is gone.

tbee committed:
improved the tests
equals requires a hashcode

tbee committed:
added equals and hashcode to the other impls as well

tbee commented on jfxtras-agenda/src/main/resources/jfxtras/internal/scene/control/skin/agenda/Agenda.css:
```diff
@@ -68,30 +68,30 @@
+.group6 { -fx-background-color: #42D692; -fx-fill: #42D692; }
+.group7 { -fx-background-color: #16A765; -fx-fill: #16A765; }
+.group8 { -fx-background-color: #7BD148; -fx-fill: #7BD148; }
+.group9 { -fx-background-color: #B3DC6C; -fx-fill: #B3DC6C; }
```
Why are these fills needed?

daviddbal replied:
I want to use rectangles as icons.  Rectangles allow some styling features, like rounded edges, that are not available to Panes.  Rectangles requires fills.

tbee commented on jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java:
```diff
@@ -238,7 +241,15 @@ public void refresh() {
+	 */
+	@Override
+    public Pane getNodeForPopup(Appointment appointment) { return appointmentNodeMap.get(appointment); }
+    final private Map<Appointment, Pane> appointmentNodeMap = new WeakHashMap<>();
```
I'm a bit worried about the hashcode method on the appointments and the hashmap being used here. What if one of the properties is changed after an appointment is put in this map?

daviddbal replied:
Yes, that is a problem.  I can think of three choices:
1.  Not override equals and hashcode.
2.  Override equals only and not override hashcode.
3.  Not use a map for getNodeForPopup.

All options present problems.
1.  I need equals for unit testing.  I will have to use my own Appointment implementation for testing again.
2.  Unknown problems due to not following expected conventions.
3.  Other matching methods are less efficient than a map (e.g. scan through a list).

Can you think of other options?

tbee replied:
1. Why do you need equals for unit testing? I've written agenda and a dozen other controls without it. You could write a static SomeUtil.isEqualTo(o1, o2). That said, Appointment is just an interface, so a user may provide his own implementation with equals en hashCode, so we cannot rely on these assumptions. The question would then be: why use a hashcode based map? 
   But then equals quickly becomes a problem. Maybe the best solution would be to use System.  identityHashCode(u) as the key, not a weak map and use a deletion listener on appointments to clean up the map.
2. not an option
3. I do not think a few simple loops will hurt that much in the grand scheme of things. But as said, you cannot use equals because there may two appointments with the same content, which we do not consider being the same.

We either decide to fully base ourselves on System.identityHashCode(u) or tell the user not to implement equals and hashcode.

daviddbal replied:
I am only using the equals method for unit testing.  All considered, it is too much trouble.  I will replace it with the static isEqualsTo method as you suggested.  That will eliminate my need for it.

I am not sure which is the better option for the getNodeForPopup method.  I need to study System.identityHashCode(u) first.

tbee replied:
identityHashCode always returns the hashCode of an object as it would if hashCode were not overridden, so two objects never are equal that way.

daviddbal replied:
I have read up on System.identityHashCode(u).  I think it should be used as the key as you suggested.  The deletion listener on appointments is easy enough to do.

You have a change listener in AllAppointments already.  Can I just add to that listener or do you think a whole new listener is best?

tbee replied:
Encapsulation would suggest a separate listener, since it would be part of a different class.

daviddbal replied:
I suppose it would be in a different class, AgendaSkinTimeScale24HourAbstract?  I'll have to think about it later.  It's too late for me and I need some sleep.  We are almost half-way around the world from each other.  You are just beginning your day.  I'm at the end of mine.

tbee replied:
I know. Isn't that great? And on my other chat is a New Sealander. 
Sleep well.

tbee commented on jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:
```diff
@@ -252,7 +252,7 @@ private void setupDragging() {
 				
                 // has the client added a callback to process the change?
-                Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getAppointmentChangedCallback();
+                Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getChangedAppointmentCallback();
```
This does not seem to belong in this PR. Please rebase and remove from the PR.

tbee commented on jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/DurationDragger.java:
```diff
@@ -141,7 +141,7 @@ private void setupMouseDrag() {
 
             // has the client added a callback to process the change?
-            Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getAppointmentChangedCallback();
+            Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getChangedAppointmentCallback();
```
This does not seem to belong in this PR. Please rebase and remove from the PR.

tbee commented on jfxtras-agenda/src/main/java/jfxtras/scene/control/agenda/Agenda.java:
```diff
@@ -349,17 +349,17 @@ private void constructAppointmentGroups() {
 	public Agenda withEditAppointmentCallback(Callback<Appointment, Void> value) { setEditAppointmentCallback(value); return this; }
 
-   /** appointmentChangedCallback:
+   /** changedAppointmentCallback:
```
This does not seem to belong in this PR. Please rebase and remove from the PR.

daviddbal committed:
don't track .gitignore

daviddbal committed:
don't track .gitignore

daviddbal committed:
return .gitignore to tracking

daviddbal committed:
merged

daviddbal committed:
appointmentNodeMap uses System.identityHashCode and clean up listener

daviddbal committed:
remove unused import, fixed appointmentChangedCallbackObjectProperty name

daviddbal commented:
Tom,

I changed the appointmentNodeMap to use System.identityHashCode(a).  I added a clean-up listener also.  The feature passes simple manual testing.  I didn't add any automatic tests.

I also rebased.

Please take a look.

-David

daviddbal committed:
Merge remote-tracking branch 'upstream/8.0' into 8.0

Conflicts:
	jfxtras-
agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java

daviddbal commented:
Are the equals and hashcode methods staying or being removed?

tbee commented:
Good question. On the one hand I'm a big fan of not having code that has no purpose, on the other hand, given that user may implement their own equals, it is good to have them in place to test against. But since Agenda must render two totally identical appointments separately, I think they should be removed.

daviddbal commented:
Maybe they should be removed from Agenda, but put in a separate test-only Appointment implementation. That way some tests can be made to guarantee correct behavior in case a user implements another equals.

daviddbal committed:
fixed problem when changing start date/time can make an invalid appointment

A valid appointment can't start after it ends.  It also can't end before
it starts.  By calling callAppointmentChangedCallback on changes to
start and then end separately can result in an invalid appointment
before the second change is recorded.  This problem is avoided by
updating both changes in one call.

daviddbal commented on jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:
```diff
@@ -299,15 +299,17 @@ private void handleDrag(Agenda.Appointment appointment, LocalDateTime dragPickup
+				appointment.setEndLocalDateTime( appointment.getEndLocalDateTime().plus(duration) );
+				changed = true;
+			}
+			if (changed) layoutHelp.callAppointmentChangedCallback(appointment);
```
A valid appointment can't start after it ends.  It also can't end before it starts.  Calling callAppointmentChangedCallback on changes to start and end separately can result in an invalid appointment after the first call, but before the second takes place.  This problem is avoided by updating both changes in one call.

My appointmenChangedCallback synchronizes the appointment to a VComponent.  It throws an exception when an invalid VComponent is created.

tbee commented on jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:
```diff
@@ -333,18 +337,20 @@ else if ( (dragPickupInDayHeader && dragDropInDayBody) ) {
-				layoutHelp.callAppointmentChangedCallback(appointment);
+                changed = true;
 			}
+            if (changed) layoutHelp.callAppointmentChangedCallback(appointment);
```
Recently I never use a single line if anymore, debugging is confusing then. Used to do it al lot though, so I'll fix this.

daviddbal replied:
OK.  I'll make sure I don't use single line if in your code in the future.  I'll even think about adopting that convention too...
